app/controllers/api/vms_controller.rb

```diff
@@ -59,6 +59,13 @@ def suspend_resource(type, id = nil, _data = nil)
     end
     central_admin :suspend_resource, :suspend
 
+    def parent_choices_resource(ids)
+      parent_choices = {}
+      parent_choices[Digest::MD5.hexdigest(ids.inspect)] ||= begin
+        Rbac.filtered(VmOrTemplate.where(:ems_id => VmOrTemplate.where(:id => ids).with_ems.group_by_ems(ids)).where.not(:id => ids))
```

Comment:
```suggestion
        ems_ids = VmOrTemplate.where(:id => ids).select(:ems_id).distinct
        Rbac.filtered(VmOrTemplate.where(:ems_id => ems_ids.where.not(:id => ids))
```

- keeping it in its own variable is easier to read
- think the `group_by` isn't quite what we want
- is it possible to tell the api that we only want a few of the columns?

@lpichler Since this is fetching vms, does this belong as a sub-collection, or a search option on vms itself?

Reply:
@lpichler I need an answer to Keenan's question, I'm blocked on this, and it'd be nice to get this moving.

Reply:
so OK, based on this [comment](https://github.com/ManageIQ/manageiq/pull/19287#issuecomment-531843518)  when are doing this API.

As I stated in this [comment](https://github.com/ManageIQ/manageiq/pull/19287#issuecomment-530470263) what we need:

```ruby
  def parent_choices(ids)
    @parent_choices = {}
    @parent_choices[Digest::MD5.hexdigest(ids.inspect)] ||= begin
      ems_ids = VmOrTemplate.where(:id => ids).distinct.pluck(:ems_id).compact
      parent_item_scope = Rbac.filtered(VmOrTemplate.where(:ems_id => ems_ids).where.not(:id => ids).order(:name))

      parent_item_scope.pluck(:name, :location, :id).each_with_object({}) do |vm, memo|
        memo[vm[0] + " -- #{vm[1]}"] = vm[2]
      end
    end
  end
```

(1) Parent VM select box
which is `parent_choices(vm.id)`


(2) Available VMs
`parent_choices([Vm.first.id] + Vm.first.children)`


So we need two endpoints because we need to avoid parameter `ids` as original method have 

@d-m-u  I guess that your method `parent_choices_resource`  tends to be (1).
so If you look [here](https://github.com/ManageIQ/manageiq-ui-classic/blob/3a817248039bcd87951069ab96cacc8f7c194254/app/controllers/vm_common.rb) you see that parameter `ids` are children of the `Vm` itself.


```ruby
vm_id  # id of resource (VM).
vm = VmOrTemplate.find(vm_id)
VmOrTemplate.where(:ems_id => vm.ems_id) .where.not(:id => vm_id)

#with RBAC
Rbac.filtered(VmOrTemplate.where(:ems_id => vm.ems_id) .where.not(:id => vm_id))
```

the case 2)

```ruby
vm_id  # id of resource (VM).
vm = VmOrTemplate.find(vm_id)
VmOrTemplate.where(:ems_id => vm.ems_id) .where.not(:id => vm_id)

#with RBAC
Rbac.filtered(VmOrTemplate.where(:ems_id => vm.ems_id).where.not(:id => [id] + children.pluck(:ids) )
```

then some logic can be moved to model:
`where(:ems_id => vm.ems_id)` -> `.with_ems(vm.ems_id)`
`where.not(:id => [id] + children.pluck(:ids))` -> `.without([id] + children.pluck(:ids))`


> - keeping it in its own variable is easier to read

Ok👍 


> - I think the group_by isn't quite what we want

it was `group_by_ems` (from https://github.com/ManageIQ/manageiq/pull/19287/files#diff-fb7c3f0121965870e99fe02ec498cf65R206 ) - but I prefer name  `with_ems` or something like this.
 
> is it possible to tell the api that we only want a few of the columns?

I think you need to add code to get requested column from params and validate it(if attribute exists on model), eventually use any existing method for it(take a look to [`parameters.rb`](https://github.com/ManageIQ/manageiq-api/blob/713b792ca06baf0be0c610b91c47940539717a81/app/controllers/api/base_controller/parameters.rb)), what we have in API and also pass column for ordering

and  keep existing format of api;

Example for

`?attributes=id,name&sort_by=name&sort_order=desc`

> @lpichler Since this is fetching vms, does this belong as a sub-collection, or a search option on vms itself?

we don't need isubcollection with just need it for one VM.
abd
I think it could be `search_options`, then @kbrock  I think, do  you have such API request on mind? 
 
Option 1)
`/api/vms/111?attributes=id,name&sort_by=name&sort_order=desc&search_options=parent_choices`

Option 2)
`/api/vms/111?attributes=id,name&sort_by=name&sort_order=desc&search_options=parent_choices_without_children`

NOTE: Not sure whether ` parent_choices_without_children ` is good.